STM32: add setting for LSE drive load level

JojoS62 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

This PR is related to https://github.com/ARMmbed/mbed-os/issues/13781
It adds a setting for the LSE drive load level. 

This settings ist target hardware dependend and 
depends on the crystal that is used for LSE clock.
For low power requirements, crystals with low load capacitors can be used and
driver setting is RCC_LSEDRIVE_LOW.
For higher stablity, crystals with higher load capacitys can be used and
driver setting is RCC_LSEDRIVE_HIGH.

When the setting is not matching the hardware, the LSE oscillator may not start or have large jitter.
Symptoms are Mbed Exceptions with LSE error or RTC wakeup error, or sleep_for() call never returns.

This PR fixes also a missing ```#include "platform/mbed_error.h"``` for the error() function that was used conditionally already.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

For known targets, the setting can be added to targets.json. For custom_targets, a default is used, but the user should provide a value that matches the hardware (crystal parameter depending).

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

The defaults may be changed and the settings for existing targets in target.json need to be defined.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

A hint for this setting is necessary in the documentation, specially for custom_targets.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    Manual tests with H743 and F411 targets were successfull.

----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

JojoS62 committed:
add setting for LSE drive load level

jeromecoutant commented on targets/TARGET_STM/mbed_overrides.c:
```diff
@@ -27,10 +27,69 @@
+#   else
+#       define LSE_DRIVE_LOAD_LEVEL    RCC_LSEDRIVE_MEDIUMHIGH
+#   endif
+#endif
```
Didn't fully check... but what about:
````
#if MBED_CONF_TARGET_LSE_DRIVE_LOAD_LEVEL
#define LSE_DRIVE_LOAD_LEVEL    MBED_CONF_TARGET_LSE_DRIVE_LOAD_LEVEL
#else
	#ifdef RCC_LSE_HIGHDRIVE_MODE
	#define LSE_DRIVE_LOAD_LEVEL    RCC_LSE_HIGHDRIVE_MODE
	#else
	#define LSE_DRIVE_LOAD_LEVEL    RCC_LSEDRIVE_MEDIUMHIGH
	#endif
#endif
````

JojoS62 replied:
accepted

jeromecoutant commented on targets/TARGET_STM/mbed_overrides.c:
```diff
@@ -27,10 +27,69 @@
+#else
+    HAL_PWR_EnableBkUpAccess();
+    __HAL_RCC_LSEDRIVE_CONFIG(LSE_DRIVE_LOAD_LEVEL);
+#endif 
```
Let's propose... :
````
HAL_PWR_EnableBkUpAccess();
#ifdef  __HAL_RCC_LSEDRIVE_CONFIG
    __HAL_RCC_LSEDRIVE_CONFIG(LSE_DRIVE_LOAD_LEVEL);
#else
    HAL_RCCEx_SelectLSEMode(LSE_DRIVE_LOAD_LEVEL);
#endif
````

JojoS62 replied:
that way it looks easier to read, I will change it.

JojoS62 replied:
accepted

jeromecoutant reviewed (commented):
Could be good to add some examples in the targets.json file with some known NUCLEO boards?

JojoS62 committed:
simplify checking for F4_g2 exception

JojoS62 committed:
remove duplicate LSEDRIVE_CONFIG

JojoS62 committed:
simplify default setting

JojoS62 committed:
fix for targets with non-modifiable
transconductance

JojoS62 committed:
fix macro usage error
use unique #if defined()

JojoS62 committed:
add defaults values
set all defaults to initial low as after a reset
most STM32 eval boards use low power crystals and work with this setting

zimjjan commented on targets/TARGET_STM/mbed_overrides.c:
```diff
@@ -27,10 +27,69 @@
+// set defaults for LSE drive load level
+#if defined(LSE_CONFIG_AVAILABLE)
+
+#   if defined(MBED_CONF_TARGET_LSE_DRIVE_LOAD_LEVEL)
```
I'm not entirely sure if something changed since [1] was documented, but I think the build system will interpret the "macro_name" from target.json _without_ adding the "MBED_CONF_" prefix, which will lead to the setting never being applied. I noticed this while I was backporting this to mbedos 5 so I might be wrong.

[1] https://os.mbed.com/docs/mbed-os/v6.11/program-setup/advanced-configuration.html

---
Edited: added the comment here, as adding it to all the target.json lines would have been a pain.

jeromecoutant replied:
Hi
I am sorry, I didn't understand well the question.
I don't think there is something different between mbedos 5 and 6 about json configuration.
And no need to add this line for each target as it can be defined at the family level.

zimjjan replied:
Sorry, I didn't made my point in an understandable way:
The implementation requires `MBED_CONF_TARGET_LSE_DRIVE_LOAD_LEVEL` to be defined to set LSEDRV from the config. However, in `target.json` the actual parameter is set to  `"macro_name": "TARGET_LSE_DRIVE_LOAD_LEVEL"` which will result in the target _not_ being generated as `MBED_CONF_TARGET_LSE_DRIVE_LOAD_LEVEL`, but without the prefix `MBED_CONF_`. (See [1] above comment)

This results in the value from `target.json` (or anywhere else, where the setting is overridden from) never being used in the first place, as far as I understand.

zimjjan replied:
I'm sorry, was resolved here https://github.com/ARMmbed/mbed-os/pull/13939
Never mind. My fault.
